discretisedfield/region.py

```diff
@@ -83,28 +73,63 @@ def __init__(
         self,
         p1,
         p2,
-        ndim=3,
-        dims=["x", "y", "z"],
-        units=["m", "m", "m"],
+        dims=None,
+        units=None,
         tolerance_factor=1e-12,
     ):
 
-        if not (ndim == len(p1) == len(p2) == len(dims) == len(units)):
+        if not isinstance(p1, (tuple, list, np.ndarray)) or not isinstance(
+            p2, (tuple, list, np.ndarray)
+        ):
+            raise TypeError(
+                "p1 and p2 must be of type tuple, list, or np.ndarray. Not"
+                f" p1={type(p1)} and p2={type(p2)}."
+            )
+
+        if not all(isinstance(i, numbers.Number) for i in p1):
+            msg = "p1 can only contain elements of type numbers.Number."
+            raise TypeError(msg)
+
+        if not all(isinstance(i, numbers.Number) for i in p2):
+            msg = "p2 can only contain elements of type numbers.Number."
+            raise TypeError(msg)
+
+        ndim = len(p1)
+        if not (len(p1) == len(p2)):
             raise ValueError(
-                "The length of p1, p2, units, and dims must all be equal and equal to"
-                " ndim."
+                "The length of p1 and p2 must be the same. Not"
+                f" len(p1)={len(p1)} and len(p2)={len(p2)}."
+            )
+
+        if not isinstance(tolerance_factor, numbers.Number):
+            raise TypeError(
+                "tolerance_factor must be of type numbers.Number. Not"
+                f" tolerance_factor={type(tolerance_factor)}."
             )
 
+        # TODO: Remove for ndim != 3 functionality.
+        if ndim != 3:
+            raise NotImplementedError("Only 3D regions are supported at the moment.")
+
+        if dims is None:
+            if ndim == 3:
+                dims = ["x", "y", "z"]
+            else:
+                dims = [f"x{i}" for i in range(ndim)]
```

Comment:
Okay for now, but we have to think about defaults and conventions for n dimensions.